### PR TITLE
Add Uniswap V2 support to @0x/asset-swapper

### PR DESCRIPTION
**Pull request by xianny**
## Description

Add Uniswap V2 to asset-swapper.

## Testing instructions

Instead of writing new tests, I replaced Uniswap (V1) with V2 in existing tests.

## Types of changes

 * New feature (non-breaking change which adds functionality)

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/market_operation_utils/orders.ts [hunk 1]
```diff
@@ -203,19 +206,39 @@ function createBridgeOrder(fill: CollapsedFill, opts: CreateOrderFromPathOpts):
...
+                makerToken,
+                bridgeAddress,
+                createUniswapV2BridgeData([makerToken, takerToken]),
+            );
+            break;
+        case ERC20BridgeSource.UniswapV2Eth:
+            makerAssetData = assetDataUtils.encodeERC20BridgeAssetData(
+                makerToken,
+                bridgeAddress,
+                createUniswapV2BridgeData([makerToken, takerToken]),
```
Comment (dekz) [inline packages/asset-swapper/src/utils/market_operation_utils/orders.ts:233]: should this be `[makerToken, WETH, takerToken]`?

File: packages/asset-swapper/src/utils/market_operation_utils/types.ts [hunk 2]
```diff
@@ -28,6 +28,8 @@ export enum AggregationError {
 export enum ERC20BridgeSource {
     Native = 'Native',
     Uniswap = 'Uniswap',
+    UniswapV2 = 'UniswapV2',
+    UniswapV2Eth = 'UniswapV2Eth',
```
Comment (dekz) [inline packages/asset-swapper/src/utils/market_operation_utils/types.ts:32]:
```suggestion
    UniswapV2 = 'Uniswap_V2',
    UniswapV2Eth = 'Uniswap_V2_ETH',
```

I promised Matcha we would stick to this pattern where the first chunk before `_` is the "real" source.

cc @johnrjj

File: packages/asset-swapper/test/market_operation_utils_test.ts [hunk 3]
```diff
@@ -997,7 +1039,7 @@ describe('MarketOperationUtils tests', () => {
                 const orderFillSources = improvedOrders.map(o => o.fills.map(f => f.source));
                 expect(orderFillSources).to.deep.eq([
                     [ERC20BridgeSource.Native],
-                    [ERC20BridgeSource.Eth2Dai, ERC20BridgeSource.Uniswap],
+                    [ERC20BridgeSource.Eth2Dai, ERC20BridgeSource.UniswapV2],
```
Comment (dekz) [inline packages/asset-swapper/test/market_operation_utils_test.ts:1042]: We should add a test for the `UniswapV2Eth` source

File: packages/asset-swapper/src/utils/market_operation_utils/orders.ts (hunk 1 above)
Comment (xianny) [reply packages/asset-swapper/src/utils/market_operation_utils/orders.ts:233]: aaaah good catch, thanks!

File: packages/asset-swapper/test/dex_sampler_test.ts [hunk 4]
```diff
@@ -336,6 +369,11 @@ describe('DexSampler tests', () => {
                     expect(fillAmounts).to.deep.eq(expectedTakerFillAmounts);
                     return fillAmounts.map(a => a.times(ratesBySource[ERC20BridgeSource.Eth2Dai]).integerValue());
                 },
+                sampleSellsFromUniswapV2: (path, fillAmounts) => {
+                    expect(path).to.deep.eq([expectedMakerToken, expectedTakerToken]);
```
Comment (xianny) [inline packages/asset-swapper/test/dex_sampler_test.ts:373]: Can I get a sanity check that the path **is* maker -> taker for sell quotes, and taker -> maker for buy quotes?

File: packages/asset-swapper/src/utils/market_operation_utils/orders.ts [hunk 5]
```diff
@@ -203,19 +206,48 @@ function createBridgeOrder(fill: CollapsedFill, opts: CreateOrderFromPathOpts):
...
-            bridgeAddress,
-            createCurveBridgeData(curveAddress, fromTokenIdx, toTokenIdx, version),
-        );
-    } else {
-        makerAssetData = assetDataUtils.encodeERC20BridgeAssetData(
-            makerToken,
-            bridgeAddress,
-            createBridgeData(takerToken),
-        );
+    switch (fill.source) {
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/orders.ts:209]: :raised_hands:

File: packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts [hunk 6]
```diff
@@ -376,6 +414,16 @@ export const samplerOperations = {
...
+                } else if (source === ERC20BridgeSource.UniswapV2) {
+                    batchedOperation = samplerOperations.getUniswapV2BuyQuotes(
+                        [takerToken, makerToken],
+                        makerFillAmounts,
+                    );
+                } else if (source === ERC20BridgeSource.UniswapV2Eth) {
+                    batchedOperation = samplerOperations.getUniswapV2BuyQuotes(
+                        [takerToken, wethAddress, makerToken],
+                        makerFillAmounts,
+                    );
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:426]: taker -> maker

File: packages/asset-swapper/test/dex_sampler_test.ts [hunk 7]
```diff
@@ -255,6 +260,28 @@ describe('DexSampler tests', () => {
...
+                    expect(fillAmounts).to.deep.eq(expectedTakerFillAmounts);
+                    return expectedMakerFillAmounts;
+                },
+            });
+            const dexOrderSampler = new DexOrderSampler(sampler);
+            const [fillableAmounts] = await dexOrderSampler.executeAsync(
+                DexOrderSampler.ops.getUniswapV2SellQuotes(
+                    [expectedMakerToken, expectedTakerToken],
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/test/dex_sampler_test.ts:278]: Should be flipped.
Other places too.

File: packages/asset-swapper/test/utils/mock_sampler_contract.ts [hunk 8]
```diff
@@ -171,6 +187,18 @@ export class MockSamplerContract extends IERC20BridgeSamplerContract {
         );
     }
 
+    public sampleBuysFromUniswapV2(
+        path: string[],
+        takerAssetAmounts: BigNumber[],
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/test/utils/mock_sampler_contract.ts:192]:
```suggestion
        makerAssetAmounts: BigNumber[],
```

File: packages/contract-addresses/addresses.json [hunk 9]
```diff
@@ -20,6 +20,7 @@
         "devUtils": "0x74134cf88b21383713e096a5ecf59e297dc7f547",
         "erc20BridgeProxy": "0x8ed95d1746bf1e4dab58d8ed4724f1ef95b20db0",
         "uniswapBridge": "0x36691c4f426eb8f42f150ebde43069a31cb080ad",
+        "uniswapV2Bridge": "0x0000000000000000000000000000000000000000",
```
Comment (dekz) [inline packages/contract-addresses/addresses.json:23]:
```suggestion
        "uniswapV2Bridge": "0xdcd6011f4c6b80e470d9487f5871a0cba7c93f48",
```

Deployed https://etherscan.io/address/0xdcd6011f4c6b80e470d9487f5871a0cba7c93f48#code

File: packages/migrations/CHANGELOG.json [hunk 10]
```diff
@@ -9,6 +9,10 @@
             {
                 "note": "Add `ERC20BridgeSampler` deployment",
                 "pr": 2541
+            },
+            {
+                "note": "Added `UniswapV2Bridge` address (null)",
```
Comment (dekz) [inline packages/migrations/CHANGELOG.json:14]:
```suggestion
                "note": "Added `UniswapV2Bridge` address on mainnet",
```

File: packages/contract-addresses/addresses.json [hunk 11]
```diff
@@ -20,6 +20,7 @@
...
         "erc20BridgeSampler": "0xc0154b14cc60a6661171fdc817f759429a57e184",
```
Comment (dekz) [inline packages/contract-addresses/addresses.json:24]: TODO update this with the latest sampler with Univ2 support.

**Pull request merged**